### PR TITLE
[SYCL] Tie BoundArch to Toolchain 

### LINKED ISSUE #4662: Target triple mismatch
**Describe the bug**
When attempting to build LAMMPS with Kokkos, I am experiencing silent failures of the offload bundling, due to a mismatch in targets. The offload bundler call is:

`"/soft/llvm/bin/clang-offload-bundler" "-type=o" "-targets=host-x86_64-unknown-linux-gnu,sycl-nvptx64-nvidia-cuda" "-inputs=CMakeFiles/lmp.dir/home/joe/sources/lammps_stack/lammps/src/main.cpp.o" "-outputs=/tmp/main-335138.o,/tmp/main-967baf.o" "-unbundle" "-allow-missing-bundles"`

but later:

`"/soft/llvm/bin/clang-offload-deps" "-targets=sycl-nvptx64-nvidia-cuda--sm_80" "-outputs=/tmp/-fffb95.bc" "/tmp/-7e2d11.out" `

Note the additional (erroneous?) `--sm_80` which means nothing is found.

**To Reproduce**

I've tried to make a reproducer but have been unable, I'm afraid. So, building LAMMPS is required.

**Environment (please complete the following information):**

- OS: Linux
- Target device and vendor: Nvidia GPU
- DPC++ version: Introduced in merge fc2f897870e264a88e5d0b3e01310490ef25465a

**Additional context**

I believe @AidanBeltonS has an upcoming PR which solves this issue.

### PR DESCRIPTION
**Pull request by AidanBeltonS**
These changes implement SYCLTargetInfoList which stores both the toolchain and the bound arch. This prevents BoundArchs being incorrectly assigned when actions and dependencies are created. This sets up the possibility of having multiple BoundArchs per triple as a future development. 
This fixes an issue where spirv and nvptx swap BoundArch's, causing failure cases. #3631
This patch should also resolve #4662

Some tests were changed in sycl-offload-with-split.c where the output from the device compilation phase was being used as an input by the wrong device. 3 tests are also introduced, two check the ordering behaviour of the triples and the bounded arch. The third checks for behaviour when targetting spirv, nvptx, and amdgcn.

### CODE REVIEW

File: clang/lib/Driver/Driver.cpp [hunk 1]
```diff
@@ -4100,33 +4114,31 @@ class OffloadingActionBuilder final {
...
-            OutputType = types::TY_LLVM_IR;
-          if (SYCLDeviceOnly) {
-            if (Args.hasFlag(options::OPT_fno_sycl_use_bitcode,
-                             options::OPT_fsycl_use_bitcode, false)) {
-              auto *CompileAction =
-                  C.MakeAction<CompileJobAction>(A, types::TY_LLVM_BC);
-              A = C.MakeAction<SPIRVTranslatorJobAction>(CompileAction,
-                                                         types::TY_SPIRV);
-              continue;
+        if (!SYCLTargetInfoList.empty()) {
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:4117]: Can the conditions on 4116-4117 be merged, now that there's nothing happening beyond this branch?

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -4472,11 +4494,11 @@ class OffloadingActionBuilder final {
             // directly to the backend compilation step (aocr) or wrapper (aocx)
             Action *FPGAAOTAction;
             if (Input->getType() == types::TY_FPGA_AOCR ||
-                Input->getType() == types::TY_FPGA_AOCR_EMU)
+                Input->getType() == types::TY_FPGA_AOCR_EMU) {
               // Generate AOCX/AOCR
               FPGAAOTAction =
                   C.MakeAction<BackendCompileJobAction>(Input, FPGAOutType);
-            else if (Input->getType() == types::TY_FPGA_AOCX)
+            } else if (Input->getType() == types::TY_FPGA_AOCX)
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:4501]: Feels like an unrelated change

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -4699,15 +4718,12 @@ class OffloadingActionBuilder final {
...
-                                          Action::OFK_SYCL);
-        });
-
-        // Add deps output to linker inputs.
+      unsigned I = 0;
+      for (auto &TargetInfo : SYCLTargetInfoList) {
+        DA->registerDependentActionInfo(TargetInfo.TC, TargetInfo.BoundArch,
+                                        Action::OFK_SYCL);
         DeviceLinkerInputs[I].push_back(DA);
+        I++;
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:4726]: One could get fancy here, but I'm not sure as to what's more readable
```suggestion
        DeviceLinkerInputs[I++].push_back(DA);
```

File: clang/lib/Driver/Driver.cpp [hunk 4]
```diff
@@ -4908,11 +4960,25 @@ class OffloadingActionBuilder final {
         }
       }
 
-      DeviceLinkerInputs.resize(ToolChains.size());
-      bool GpuInitHasErrors = initializeGpuArchMap();
-      if (ShouldAddDefaultTriple)
-        addSYCLDefaultTriple(C, SYCLTripleList);
-      return GpuInitHasErrors;
+      if (ShouldAddDefaultTriple) {
+        if (addSYCLDefaultTriple(C, SYCLTripleList)) {
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:4964]: The defined order of condition evaluation should allow to reduce the nesting:
```suggestion
      if (ShouldAddDefaultTriple && addSYCLDefaultTriple(C, SYCLTripleList)) {
```

File: clang/lib/Driver/Driver.cpp [hunk 5]
```diff
@@ -6947,12 +7013,12 @@ InputInfo Driver::BuildJobsForActionNoCache(
...
     }
   } else {
-    if (UnbundlingResults.empty())
+    if (UnbundlingResults.empty()) {
       T->ConstructJob(
           C, *JA, Result, InputInfos,
           C.getArgsForToolChain(TC, BoundArch, JA->getOffloadingDeviceKind()),
           LinkingOutput);
-    else
+    } else
```
Comment (AGindinson) [inline clang/lib/Driver/Driver.cpp:7021]: Feels like an unrelated change

File: clang/test/Driver/sycl-offload-with-split.c [hunk 6]
```diff
@@ -245,7 +245,7 @@
 // CHK-PHASE-MULTI-TARG: 3: input, "[[INPUT]]", c++, (device-sycl)
 // CHK-PHASE-MULTI-TARG: 4: preprocessor, {3}, c++-cpp-output, (device-sycl)
 // CHK-PHASE-MULTI-TARG: 5: compiler, {4}, ir, (device-sycl)
-// CHK-PHASE-MULTI-TARG: 6: offload, "host-sycl (x86_64-unknown-linux-gnu)" {2}, "device-sycl (spir64-unknown-unknown)" {5}, c++-cpp-output
+// CHK-PHASE-MULTI-TARG: 6: offload, "host-sycl (x86_64-unknown-linux-gnu)" {2}, "device-sycl (spir64_gen-unknown-unknown)" {5}, c++-cpp-output
```
Comment (AGindinson) [inline clang/test/Driver/sycl-offload-with-split.c:248]: Note to self & other reviewers: this change concerned me at first, as I've doubted the deterministic ordering of the target toolchains after the map introduction, but then realized that usage of `MapVector` for toolchain/arch mapping [ensures that](https://llvm.org/doxygen/classllvm_1_1MapVector.html#details).

**Review by AGindinson — COMMENTED** [review]
The change looks good conceptually (although @mdtoguchi would have the final say here). The bulk of my comments is style-related - in case any serious concerns arise, I'm hoping to roll all of them out in the next iteration after fully grasping the changes.

File: clang/lib/Driver/Driver.cpp (hunk 1 above)
Comment (AidanBeltonS) [reply clang/lib/Driver/Driver.cpp]: Yes, the conditions can be merged. I have made the change.

File: clang/lib/Driver/Driver.cpp [hunk 7]
```diff
@@ -681,21 +682,22 @@ static bool isValidSYCLTriple(llvm::Triple T) {
   return true;
 }
 
-static void addSYCLDefaultTriple(Compilation &C,
+static bool addSYCLDefaultTriple(Compilation &C,
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply clang/lib/Driver/Driver.cpp:685]: I have added a comment `/// Returns true if a triple is added to SYCLTriples, false otherwise`

**Review by AGindinson — APPROVED** [review]
Thanks for addressing my comments - LGTM! Approving right away to unblock the checks, yet I'd prefer a second approval from the component owners.

File: clang/test/Driver/sycl-offload.cpp [hunk 8]
```diff
@@ -68,7 +68,7 @@
 // RUN:    | FileCheck -check-prefix IMPLIED_DEVICE_OBJ %s
 // RUN:  %clangxx -### -target x86_64-unknown-linux-gnu -fsycl -fintelfpga %S/Inputs/SYCL/objlin64.o %s 2>&1 \
 // RUN:    | FileCheck -check-prefix IMPLIED_DEVICE_OBJ %s
-// IMPLIED_DEVICE_OBJ: clang-offload-bundler{{.*}} "-type=o"{{.*}} "-targets={{.*}}sycl-spir64-unknown-unknown,sycl-spir64_{{.*}}-unknown-unknown"{{.*}} "-unbundle"
+// IMPLIED_DEVICE_OBJ: clang-offload-bundler{{.*}} "-type=o"{{.*}} "-targets=sycl-spir64_{{.*}}-unknown-unknown,{{.*}}sycl-spir64-unknown-unknown"{{.*}} "-unbundle"
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-offload.cpp:71]: Looks like the ordering swap isn't taking into account the host triple which is coming in at the front of the `-targets`

Comment (AidanBeltonS) [reply clang/test/Driver/sycl-offload.cpp]: Thanks! I have added the host triple to the start of the expression. It was only failing in windows as if there are device libraries to link with then there is a `clang-offload-bundler` which does not have the host triple but does have both device triples.

**Review by AGindinson — APPROVED** [review]

**Pull request merged**